### PR TITLE
cgroups: fs: fix cgroup.Parent path sanitisation

### PR DESCRIPTION
**Pull request by cyphar**
Properly sanitise the --cgroup-parent path, to avoid potential issues
(as it starts creating directories and writing to files as root). In
addition, fix an infinite recursion due to incomplete base cases.

Signed-off-by: Aleksa Sarai <asarai@suse.com>

/cc @security-team @diogomonica

### CODE REVIEW

File: libcontainer/cgroups/fs/cpuset.go [hunk 1]
```diff
@@ -95,6 +95,10 @@ func (s *CpusetGroup) ensureParent(current, root string) error {
 	if filepath.Clean(parent) == root {
 		return nil
 	}
+	// Avoid infinite recursion.
+	if parent == current {
+		return nil
+	}
```
Comment (hqhq) [inline libcontainer/cgroups/fs/cpuset.go]: In what case would parent equals to current? Did you hit this in real life?

Comment (cyphar) [inline libcontainer/cgroups/fs/cpuset.go]: Yes. It causes Docker to crash in not-very-fun ways.

**Comment by dqminh** [comment]
Mmm i think we should have test for edge case like this, maybe eventually runc should test itself instead of depending on the docker test suite.

For now, can we add a test on the corresponding vendor update in docker ?

**Comment by cyphar** [comment]
@dqminh I'll include it in the vendor patch now.

File: libcontainer/cgroups/fs/cpuset.go (hunk 1 above)
Comment (hqhq) [inline libcontainer/cgroups/fs/cpuset.go]: Can you be more specific, like how can we reproduce this with docker? I still don't know how could `filepath.Dir(path) == path` and it's not returned by "filepath.Clean(parent) == root".

And I think it'll help to add more comments for such code.

Comment (cyphar) [inline libcontainer/cgroups/fs/cpuset.go]: The case `--cgroup-parent=../../../../../hello/this/is/a/crash` will cause Docker to crash due to an infinite recursion error. While this is solved with my path cleaning code, it implies that there are base cases not accounted for. `root` isn't `/`, it's just a regular path. So if `path` is a not a subdirectory of `root`, this will break. I can make it panic (or return an error) if we ever hit the case that `root` is a subdirectory of `path`.

**Comment by cyphar** [comment]
@hqhq I've made it emit an error (and have a slightly more descriptive comment). I don't think it makes sense to omit this check, even though the parent path stuff fixes the root cause (this incomplete base case causes lots of versions of Docker to be vulnerable to a DoS).

File: libcontainer/cgroups/fs/apply_raw.go [hunk 2]
```diff
@@ -230,12 +230,25 @@ func (m *Manager) GetPids() ([]int, error) {
...
 
+// absClean cleans a path such that it can be safely appended to any path and
+// the resulting path will not escape the prefixed path. This is done entirely
+// lexically, with no filesystem calls. Thus, it does not account for symlinks
+// (which can cause security problems).
+func absClean(path string) string {
+	path = filepath.Clean(string(os.PathSeparator) + path)
+	path, _ = filepath.Rel(string(os.PathSeparator), path)
+	return path
+}
```
Comment (hqhq) [inline libcontainer/cgroups/fs/apply_raw.go]: If `path = /a/b/c`, `absClean(path)` will end up with `a/b/c`, seems not correct.

Comment (cyphar) [inline libcontainer/cgroups/fs/apply_raw.go]: Wait what? This isn't the latest version of the commit. One sec.

**Comment by cyphar** [comment]
Sorry for the wrong version of the patch. Check it again.

/ping @dqminh @mrunalp @vishh @crosbymichael @hqhq

**Comment by hqhq** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM
